### ISSUE TITLE
Add ability to move inner functions

### ISSUE THREAD

**Issue by lmmx** [lmmx/mvdef#18]
E.g. this repo https://github.com/test123/Sorting-Algorithm-Visualizer uses inner functions in `sorting.py`, such as `heapify` (within `heap_sort`)

```py
def heap_sort(nums):  # n * logn
    def heapify(nums, heap_size, root_index):
        # Assume the index of the largest element is the root index
        largest = root_index
        left_child = (2 * root_index) + 1
        right_child = (2 * root_index) + 2
```

Running `mvdef -m bubble_sort sorting.py bubble.py` to refactor out the bubble sort function into its own module raises an error:

```py
ValueError ⠶ These names could not be sourced: ['root_index', 'heap_size', 'heapify']
```

in full:

```py
mvdef -m bubble_sort sorting.py bubble.py --show-tracebacks
Traceback (most recent call last):
  File "/home/louis/miniconda/envs/mvdef-edit-test/bin/mvdef", line 33, in <module>
    sys.exit(load_entry_point('mvdef', 'console_scripts', 'mvdef')())
  File "/home/louis/dev/mvdef/src/mvdef/__main__.py", line 79, in main
    raise e
  File "/home/louis/dev/mvdef/src/mvdef/__main__.py", line 74, in main
    run_cli(src_path, dst_path, mvdefs, dry_run, report, backup)
  File "/home/louis/dev/mvdef/src/mvdef/cli.py", line 10, in main
    src_parsed, dst_parsed = parse_transfer(
  File "/home/louis/dev/mvdef/src/mvdef/transfer.py", line 400, in parse_transfer
    raise link.src.edits
  File "/home/louis/dev/mvdef/src/mvdef/transfer.py", line 289, in __init__
    self.src.ast_parse()  # populate self.src.edits
  File "/home/louis/dev/mvdef/src/mvdef/transfer.py", line 93, in ast_parse
    self.retrieve_ast_agenda(transfers)  # parse AST, populate the edits property
  File "/home/louis/dev/mvdef/src/mvdef/ast_util.py", line 52, in retrieve_ast_agenda
    linkfile.edits = linkfile.process_ast(trunk, transfers)
  File "/home/louis/dev/mvdef/src/mvdef/ast_util.py", line 112, in process_ast
    m_names, nm_names, rm_names = parse_mv_funcs(linkfile, trunk)
  File "/home/louis/dev/mvdef/src/mvdef/ast_util.py", line 422, in parse_mv_funcs
    nonmvdef_names = get_def_names(
  File "/home/louis/dev/mvdef/src/mvdef/ast_util.py", line 344, in get_def_names
    raise ValueError(f"These names could not be sourced: {unknowns}")
ValueError: These names could not be sourced: ['heapify', 'root_index', 'heap_size']
```

**Comment by lmmx** [comment]
Likewise this repo https://github.com/zamhown/sorting-visualizer can be run more simply, with `python output.py play`, and again it has subfunction `def`s (`animate`) in `output.py`, which raise an error when trying to move any defs from that file.

**Comment by lmmx** [comment]
Inner functions are now annotated as `FuncDef.inner_funcs` but this isn't used yet.

https://github.com/lmmx/mvdef/blob/8f1289f0eabeb623de7fc33d9538a0a1dd32b1fe/src/mvdef/ast_util.py#L513

Should have a 'path' into the funcdef, so that two functions can share a name but cannot share a path (and treat the path as its name for the purpose of indexing and then convert back at the final step?)

**Comment by lmmx** [comment]
- `extradef_names` refers to names which are encountered while walking outside of both import and funcdef nodes, and thus this is not relevant to inner funcs which are by definition within funcdef nodes (so will never be `extra`)
- `intradef_names` can refer to (for now) a list of the names of the inner funcdefs, and this will conveniently be expandable into a dict (as iteration over a dict gives its keys, so easy to add the features needed from values by changing the type later)
  - the values can be the path of names [parent_funcdef:inner_funcdef], and the keys can be the names (for the purpose of part 1 only the names are needed)

**Comment by lmmx** [comment]
Made an attempt to move inner functions but nothing is moved. TODO: debug

**Comment by lmmx** [comment]
If I had to guess I think it’s perhaps this line?

https://github.com/lmmx/mvdef/blob/befa8f2901953663f04e19422674234ac30ac6b8/src/mvdef/ast_util.py#L407

May have to go back to the drawing board and separate out more clearly the way the inner funcdefs were excluded when processing the top-level defs so as to avoid excluding them (contradictorily) when processing the inner funcdefs...

**Comment by lmmx** [comment]
I think that what's going on here is that I need all the variable IDs (i.e. string names) which denote funcdefs available to the namespace, and when the funcdef `fd` is an innerfunc (as captured in the `if ":" in m` conditional block) then this namespace will become a recursive namespace down to the level of the innerfunc `fd`, and as such requiring all namespace IDs to be provided. This is best provided through a `InnerFuncDef` class property (which will not be set for a regular `FuncDef`, or I could make it consistent and set the current value of `fd_ids` through there but I don't think that would make sense, i.e. have little semantic meaning in and of itself: this is a computation needed for pragmatic reasons only)

**Comment by lmmx** [comment]
While programming this I realised there's going to be a race condition wherein the `ast_util.FuncDef` class `__init__` method involves `check_for_inner_funcs` [after `super().__init__` --> `ast.FunctionDef.__init__`] during which a listcomp is carried out to create `InnerFuncDef` and set these on the `FuncDef.inner_funcs` property, **but** upon initialisation these will in turn immediately call `super().__init__` --> `ast_util.FuncDef.__init__` and then `check_for_inner_funcs` at which point there will be a failed race condition to access the `def_name`...

In short, the inner func of a global-scope funcdef will be fine, but an inner func of an inner func will involve an `__init__` procedure which cannot succeed, as the required properties/attributes are not set. This race condition can be avoided by not running `check_for_inner_funcs` if the funcdef itself is an inner func

- It's not immediately clear if this can be relaxed to: not running `check_for_inner_funcs` if the funcdef parent is an inner func

**Comment by lmmx** [comment]
Achieved recursive creation without race condition https://github.com/lmmx/mvdef/blob/3fefa038ae5d292142ee1f453f732f14e2c66ebb/src/mvdef/ast_util.py#L457

TODO: uncomment out the `ns_fd_ids` and handle this https://github.com/lmmx/mvdef/blob/3fefa038ae5d292142ee1f453f732f14e2c66ebb/src/mvdef/ast_util.py#L392

**Comment by lmmx** [comment]
The inner function is now successfully addressed in `ast_util.py` (https://github.com/lmmx/mvdef/commit/11e28cde59cd003778520cac7a236554df0ce483)

but going from `SrcFile.mvdefs` --> `SrcFile.defs_to_move` requires "passage" through `ast_tokens.py`⠶`get_defs` (as the `copy_src_defs_to_dst` function iterates over the `defs_to_move` property and creates the underlying attribute as it does so 'on the fly')

- Note that this seems unnecessary but recall that the `ast_tokens` module provides some annotation not available through the standard library `ast` module which was used up to this point, so it is actually necessary to recompute this step.

As `get_defs`

**Comment by lmmx** [comment]
Now that it’s clear that implementing this will require storing an index for all inner functions, I’m wondering whether I should also implement (via the same mechanism) an index for all top-level/global scope functions (`LinkedFile.ast_defs`).

A benefit of unifying this would mean it’d make it clearer how to have a common interface for classes and their methods

**Comment by lmmx** [comment]
This can now be used as:

```sh
python -i def_path_util.py
```
⇣
```py
>>> p = FuncDefPathString("Foo::Bar.baz@setter")
>>> for part in p.parts: print(f"{part=} ({part.part_type})")
... 
part='Foo' (Class)
part='Bar' (InnerClass)
part='baz' (Method)
part='setter' (Decorator)
```

**Comment by lmmx** [comment]
As just noted in a comment, the use cases where this fails can’t really be guessed at this stage, so I’ll implement if/when I find them (I suspect something like a conditional block which ‘gets in the way’ of a direct descendant down to the targeted funcdef node, maybe)

**Comment by lmmx** [comment]
The next step will be editing this part https://github.com/lmmx/mvdef/blob/54a1f2b56b415d9803af000cb9d42a9d9d4965f1/src/mvdef/ast_util.py#L437-L445

For now this runs on the assumption that the top level will be a funcdef, but it’d actually be trivial to add in support for top level classdefs

You’d just either need an equivalent `linkfile.ast_clsdefs` to `linkfile.ast_defs` or to make one on the fly... but probably best to make it a property constructed at the same point as `ast_defs` (TODO: where is this exactly?)

```py
if len(m_parsed.parts) > 1:
    # Multi-part path, separated by one or more of the following separators
    # `:` (inner func), `.` (method), `::` (inner class), `@` (decorator)
    m_path = InnerFuncPath(m) # make fresh subclass of FuncDefPathString
    # InnerFuncPath will error if `m` does not begin with 2 funcdefs
    fd = m_path.check_against_linkedfile(linkfile) # retrieve funcdef from AST
    fd_ids = fd.all_ns_fd_ids # inner funcdef IDs, includes global def namespace
elif m in [f.name for f in linkfile.ast_defs]:
     fd = linkfile.ast_defs[[f.name for f in linkfile.ast_defs].index(m)]
 ```

**Comment by lmmx** [comment]
Edited the above to initially support only `InnerFunc`, but in future can expand this to support classes: https://github.com/lmmx/mvdef/blob/b678935f5bfa7450a30bc515aeb8e158dd472f64/src/mvdef/ast_util.py#L427-L437

Additionally, expanded the section of `asttokens.py` so `defs_to_move` will now be populated with multi-part path funcdefs: https://github.com/lmmx/mvdef/blob/b678935f5bfa7450a30bc515aeb8e158dd472f64/src/mvdef/ast_tokens.py#L47-L78

**Comment by lmmx** [comment]
Now that `defs_to_move` is populated for inner function funcdefs, we can get the following trivial result:

- `hello.py`

```py
def world():
    def foo():
        def bar():
            def baz():
                def woo():
                    print("blah")
                def zoo():
                    pass
                print("blah")
            def zaz():
                pass
            print("blah")
        def zar():
            pass
        print("blah")
    def zoo():
        print("hello")
    zoo()

def hello():
    print("hi")
```

- `world.py`

```py
from pprint import pprint
def myfunc():
    return "foo"


def hello_pretty():
    x = {0 for _ in range(3)}
    pprint("hi")
```

Running `mvdef -m world:foo:bar hello.py world.py` moves the inner function `bar` from the `world:foo` scope of `hello.py` and simply dumps its lines onto the end of `world.py` (without changing the indentation!)

- `hello.py`

```py
def world():
    def foo():
        def zar():
            pass
        print("blah")
    def zoo():
        print("hello")
    zoo()

def hello():
    print("hi")
```

- `world.py`

```py
from pprint import pprint
def myfunc():
    return "foo"


def hello_pretty():
    x = {0 for _ in range(3)}
    pprint("hi")


        def bar():
            def baz():
                def woo():
                    print("blah")
                def zoo():
                    pass
                print("blah")
            def zaz():
                pass
            print("blah")
```

For now this is a bug, and needs to be fixed by handling the inner function move more carefully...

**Comment by lmmx** [comment]
Thanks to https://github.com/lmmx/mvdef/commit/f4775c8ee1ff4033a5a815ab8a505bc35873c1e1, the above example will now run to give the following output for `world.py`, showing successful ability to move inner functions:

```py
from pprint import pprint
def myfunc():
    return "foo"


def hello_pretty():
    x = {0 for _ in range(3)}
    pprint("hi")


def bar():
    def baz():
        def woo():
            print("blah")
        def zoo():
            pass
        print("blah")
    def zaz():
        pass
    print("blah")
```
